walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,188 @@
 
 package com.google.sps;
 
+import com.google.sps.TimeRange;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  public static final int WHOLE_DAY = 24 * 60;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> meetings = new ArrayList<>();
+    Collection<String> requestedAttendees = request.getAttendees();
+    Collection<String> requestedOptionalAttendees = 
+      request.getOptionalAttendees();
+    long durationOfRequest = request.getDuration();
+
+    // Case 0: if duration too long (longer than a day), return no options
+    if (durationOfRequest > WHOLE_DAY) {
+      return meetings;
+    }
+
+    /** Case 1: GENERAL CASE >> remove conflicting times(from @events) 
+      * from possible meetings
+      */
+    meetings.add(TimeRange.fromStartDuration(0, WHOLE_DAY));
+    ArrayList<Event> optionalOverlappedEvents = new ArrayList<>();
+    for (Event event : events) {
+      Set<String> tempOverlappedEventAttendees = event.getAttendees();
+
+      Set<String> overlappedEventAttendees = new HashSet<String>();
+      Set<String> optionalOverlappedAttendees = new HashSet<String>();
+
+      // Copy attendees into a modifiable collection
+      overlappedEventAttendees = 
+        copyToModifiableCollection(
+          overlappedEventAttendees, 
+          tempOverlappedEventAttendees);
+      // Copy optional attendees into a modifiable collection
+      optionalOverlappedAttendees = 
+        copyToModifiableCollection(
+          optionalOverlappedAttendees, 
+          tempOverlappedEventAttendees);
+
+      // Filter events for overlaps with request
+      overlappedEventAttendees.retainAll(requestedAttendees);
+      optionalOverlappedAttendees.retainAll(requestedOptionalAttendees);
+
+      if (overlappedEventAttendees.size() != 0) {
```

Comment:
better to use isEmtpy() as the function captures what you are doing with the logic and it is more readable